components/core/src/flowcontrol/backoff.rs

```diff
@@ -0,0 +1,84 @@
+use std::time::Duration;
+
+use rand::{distributions::Uniform,
+           prelude::Distribution,
+           thread_rng};
+use tokio::time::Instant;
+
+#[derive(Debug, Clone)]
+pub struct Backoff {
+    base_backoff: Duration,
+    max_backoff:  Duration,
+    multiplier:   f64,
+    last_attempt: Option<RetryAttempt>,
+}
+
+#[derive(Debug, Clone)]
+struct RetryAttempt {
+    attempted_at:   Instant,
+    sleep_duration: Duration,
+}
+
+impl Default for Backoff {
+    fn default() -> Self {
+        Self { base_backoff: Duration::from_secs(10),
+               max_backoff:  Duration::from_secs(300),
+               multiplier:   3f64,
+               last_attempt: None, }
+    }
+}
+
+impl Backoff {
+    pub fn new(base_backoff: Duration, max_backoff: Duration, multiplier: f64) -> Backoff {
+        Backoff { base_backoff,
+                  max_backoff,
+                  multiplier,
+                  last_attempt: None }
+    }
+
+    // Record the completion of attempting an operation
+    pub fn record_attempt(&mut self) -> Option<Duration> {
+        match &self.last_attempt {
+            Some(last_attempt)
+                if Instant::now().duration_since(last_attempt.attempted_at)
+                   >= last_attempt.sleep_duration =>
+            {
+                let distribution = Uniform::new(self.base_backoff,
```

Comment:
probably not a huge deal but the [docs](https://docs.rs/rand/0.8.5/rand/distributions/uniform/struct.Uniform.html) on `Uniform` suggest using [`Rng::gen_range`](https://docs.rs/rand/0.8.5/rand/trait.Rng.html#method.gen_range) when only using a single sample to avoid the upfront cost of `Uniform`.

Reply:
Thanks, yes that would be a better function to use here.